### ISSUE TITLE
Websockets issue on Vaadin application with custom url mapping

### ISSUE THREAD

**Issue by onuridrisoglu** [vaadin/flow#15565]
### Description of the bug

After migrating to v23.3.1, the web sockets doesn't work anymore in our application. 
Seems like related with this change: https://github.com/vaadin/flow/pull/15188

We have two mappings for VaadinServlet like below:
```
        <servlet-mapping>
            <servlet-name>Portal</servlet-name>
            <url-pattern>/app/*</url-pattern>
        </servlet-mapping>
        <servlet-mapping>
            <servlet-name>Portal</servlet-name>
            <url-pattern>/VAADIN/*</url-pattern>
        </servlet-mapping>
```

and the web socket endpoint is set incorrectly:

```
[INFO    ] CWWKH0046I: Adding a WebSocket ServerEndpoint with the following URI: /VAADIN/VAADIN/push
[INFO    ] CWWKH0046I: Adding a WebSocket ServerEndpoint with the following URI: /VAADIN/VAADIN/push/
```

PS: In version 22.2.11 everything works fine.

### Expected behavior

Websockets should work as expected.

### Minimal reproducible example

Just downloaded a skeleton project, and added the a web.xml file with following content:

```
<servlet>
            <servlet-name>Portal</servlet-name>
            <servlet-class>com.vaadin.flow.server.VaadinServlet</servlet-class>
            <load-on-startup>1</load-on-startup>
            <async-supported>true</async-supported>
        </servlet>
        <servlet-mapping>
            <servlet-name>Portal</servlet-name>
            <url-pattern>/app/*</url-pattern>
        </servlet-mapping>
        <servlet-mapping>
            <servlet-name>Portal</servlet-name>
            <url-pattern>/VAADIN/*</url-pattern>
        </servlet-mapping>
```

### Versions

- Vaadin / Flow version: v23.3.1
- Java version: Java 11
- OS version: Mac OS
- Browser version (if applicable): Chrome Version 108.0.5359.124
- Application Server (if applicable): Open Liberty 22.0.0.13

**Comment by craigwebb-xft** [comment]
We are having the same problem. The reason is that in PushRequestHandler.initAtmosphere(ServletConfig) when setting the InitParameter for JSR356_MAPPING_PATH it calls findFirstUrlMapping()  which sorts the mappings, and returns the first one.

@onuridrisoglu has 2 mappings - /app/* and /VAADIN/* - and because /VAADIN comes before /app when sorted, /VAADIN/ is prepended to the Push URL, hence /VAADIN/VAADIN/push, and I assume that /app/ is actually the URL that needs prepending.

We also have multiple mappings, and it is picking the wrong one, but currently no way round it. Could we set an InitParameter for the PushURLPrefix, and the PushRequestHandler can check for that when setting the JSR356_MAPPING_PATH?

**Comment by tepi** [comment]
@craigwebb-xft I can confirm that this is broken, but will need some more details to come up with a reasonable way forward. 

May I ask what is the use case for the multiple mappings? The example @onuridrisoglu posted in the ticket description will just have the VaadinServlet serve everything from both /app/* and /VAADIN/*, so it's not clear why two mappings for the same VaadinServlet are needed in the first place. Do you have some custom handling in place for the separate mappings since Vaadin itself just seems to serve the same application from both mappings.

**Comment by craigwebb-xft** [comment]
@tepi We use multiple UI frameworks in our project including Vaadin 7/8. Because Vaadin 7/8 and Vaadin Flow use /VAADIN/ we need to use custom handling to make Vaadin 7/8 and Flow work together.

**Comment by craigwebb-xft** [comment]
@tepi Is there any update on this please? We would like to start using TabSheet which is only available in 23.3, but we are stuck on 23.2 until this issue is resolved.

**Comment by Artur-** [comment]
Can't we just skip `/VAADIN/*` as it is never the mapping we should use for push?

**Comment by tepi** [comment]
We can easily skip the /VAADIN/* mapping, but there can still be multiple mappings like /admin/*, /app/* and so on. We can only have the ws endpoint at one url, and currently we do not know on client-side which is the "first" mapping. We can't use serviceUrl either since that seems to change on client-side depending on which mapping you navigate to. So I think we need to reintroduce the pushUrl param in some form (maybe only as an internal feature). What do you think @Artur- ?

In the case above we would then have ws mapped to /admin/VAADIN/push, even when /app is visited.

**Comment by Artur-** [comment]
I think we first should resolve this particular case with `/VAADIN/*` by ignoring it. The case with multiple servlets and push is not new and it is not even 100% clear that a push URL of `/admin/VAADIN/push` for `/app` is a problem

**Comment by tepi** [comment]
Alright, I'll do a simple change to ignore the `/VAADIN/*` mapping, and do another PR for the more complex case.

**Comment by tepi** [comment]
After some consideration, the planned fix to map the WS endpoint (for multiple servlet mappings) like this:

1. if a parameter called `pushServletMapping` is defined, that will be used
2. otherwise the WS endpoint will be mapped to the first servlet mapping, ignoring `/VAADIN/*` mapping

This should allow the following to happen:
- WS endpoint will never be mapped under `/VAADIN/*` mapping
- WS push should work for multiple mappings automatically via the first mapping
- If needed, developers can define a specific mapping to use for the WS endpoint, which should then be used regardless of which mapping the browser is navigated to

@craigwebb-xft @onuridrisoglu Would this fix the issue for you?

**Comment by craigwebb-xft** [comment]
@tepi Yes that will work for us thanks. Will the `pushServletMapping` be the full URL (`/flow/VAADIN/push`) or just the prefix (`/flow/`)? Either will work, but probably easiest to just set the prefix, like we did with the `pushUrl` before it was removed.

**Comment by craigwebb-xft** [comment]
Could this fix be applied to Vaadin 23 as well please?

**Comment by mshabarov** [comment]
This fix has been picked into Flow 23.3 branch, so should be available in the next Vaadin 23.3.x maintenance release.
